test/hotspot/jtreg/serviceability/HeapDump/FieldsInInstanceTest.java

```diff
@@ -0,0 +1,153 @@
+import jdk.test.lib.hprof.model.Snapshot;
+import jdk.test.lib.hprof.parser.Reader;
+
+/*
+ * @test
+ * @bug 8317692
+ * @summary Verifies heap dump contains all fields of an instance
+ * @library /test/lib
+ * @run driver FieldsInInstanceTest
+ */
+class FieldsInInstanceTarg extends LingeredApp {
+
+    public static void main(String[] args) {
+        B b = new B();
+        LingeredApp.main(args);
+        Reference.reachabilityFence(b);
+    }
+
+    interface I {
+            int i = -10;
+    }
+    static abstract class A implements I {
+        static boolean b;
+        int a = 3;
+        String s = "Field";
+    }
+    static class B extends A {
+        static String f = null;
+        int a = 7;
+        double s = 0.5d;
+    }
+}
+
+public class FieldsInInstanceTest {
+
+    public static void main(String[] args) throws Exception {
+        File dumpFile = new File("Myheapdump.hprof");
+        createDump(dumpFile, args);
+        verifyDump(dumpFile);
+    }
+
+    private static void createDump(File dumpFile, String[] extraOptions) throws Exception {
+        LingeredApp theApp = null;
+        try {
+            theApp = new FieldsInInstanceTarg();
+
+            List<String> extraVMArgs = new ArrayList<>();
+            extraVMArgs.addAll(Arrays.asList(extraOptions));
+            LingeredApp.startApp(theApp, extraVMArgs.toArray(new String[0]));
+
+            //jcmd <pid> GC.heap_dump <file_path>
+            JDKToolLauncher launcher = JDKToolLauncher
+                    .createUsingTestJDK("jcmd")
+                    .addToolArg(Long.toString(theApp.getPid()))
+                    .addToolArg("GC.heap_dump")
+                    .addToolArg(dumpFile.getAbsolutePath());
+            Process p = ProcessTools.startProcess("jcmd", new ProcessBuilder(launcher.getCommand()));
+            // If something goes wrong with heap dumping most likely we'll get crash of the target VM.
+            while (!p.waitFor(5, TimeUnit.SECONDS)) {
+                if (!theApp.getProcess().isAlive()) {
+                    log("ERROR: target VM died, killing jcmd...");
+                    p.destroyForcibly();
+                    throw new Exception("Target VM died");
+                }
+            }
+
+            if (p.exitValue() != 0) {
+                throw new Exception("Jcmd exited with code " + p.exitValue());
+            }
+        } finally {
+            LingeredApp.stopApp(theApp);
+        }
+    }
+
+    private static void verifyDump(File dumpFile) throws Exception {
+        Asserts.assertTrue(dumpFile.exists(), "Heap dump file not found.");
+
+        log("Reading " + dumpFile + "...");
+        try (Snapshot snapshot = Reader.readFile(dumpFile.getPath(), true, 0)) {
+            log("Resolving snapshot...");
+            snapshot.resolve(true);
+            log("Snapshot resolved.");
+
+            Iterable<JavaHeapObject> objects = snapshot.getThings()::asIterator;
+            for (JavaHeapObject heapObj : objects) {
```

Comment:
Instead if iteration over all dumped objects it would be faster to:
`
       (JavaObject)snapshot.findClass(className).getInstances(false).nextElement();
`